### PR TITLE
Gen3: fixes I2C hang up issue

### PR DESCRIPTION
**Pull request by XuGuohui**
### Problem

I2C might get hang up by slave device and never lease the bus.

### Solution

Reset I2C bus by simulating 9 clocks on the SCL line and reset the I2C state machine.

This PR also fixes the issue that Fuel Gauge returns 256 if it fails to read the SoC.

### Steps to Test

1. Build and flash the given example below for B5SoM or BSoM (SoM evaluation board is required).

2. Connect the SoM evaluation board to your computer and open a serial monitor to watch the log.

3. Disconnect and reconnect the `PM_SCL` jumper on the SoM evaluation board randomly, quickly and repeatedly.

4. See if the SoC is get updated as long as the `PM_SCL` jumper is connected. And if the `PM_SCL` is disconnected, the SoC should be `-1`.

### Example App

```c
#include "Particle.h"

SYSTEM_MODE(MANUAL);

SerialLogHandler log(115200, LOG_LEVEL_ALL);

FuelGauge fuel;

void setup() {
}

void loop() {
    static system_tick_t startTime = millis();
    if (millis() - startTime >= 500) {
        startTime = millis();
        LOG(INFO, "SoC: %4.2f %%", fuel.getSoC());
    }
}
```

### Completeness

- [x] User is totes amazing for contributing!
- [x] Contributor has signed CLA ([Info here](https://github.com/spark/firmware/blob/develop/CONTRIBUTING.md))
- [ ] Problem and Solution clearly stated
- [ ] Run unit/integration/application tests on device
- [ ] Added documentation
- [ ] Added to CHANGELOG.md after merging (add links to docs and issues)

### CODE REVIEW

File: hal/src/nRF52840/i2c_hal.cpp [hunk 1]
```diff
@@ -391,19 +390,20 @@ int32_t HAL_I2C_Request_Data_Ex(HAL_I2C_Interface i2c, const HAL_I2C_Transmissio
         // FIXME: There is a bug in nrfx_twim driver, if we call nrfx_twim_rx repeatedly and quickly,
         // p_cb->busy will be set and never cleared, in this case nrfx_twim_rx always returns busy error
         LOG_DEBUG(TRACE, "BUSY ERROR, restore twi.");
-        twi_uninit(i2c);
```
Comment (avtolstoy) [inline hal/src/nRF52840/i2c_hal.cpp:394]: Do we still need to reinitialize I2C here?

File: wiring/src/spark_wiring_fuel.cpp [hunk 2]
```diff
@@ -119,6 +119,9 @@ float FuelGauge::getSoC() {
 	byte LSB = 0;
 
 	readRegister(SOC_REGISTER, MSB, LSB);
+    if (MSB > 99 || LSB == 255) {
```
Comment (avtolstoy) [inline wiring/src/spark_wiring_fuel.cpp:122]: Nit: formatting (spaces/tabs).

Also, this does not seem like a very reliable approach to track errors. If we are trying to fix this, `readRegister()` should be checking for errors and propagating them as well.

Comment (XuGuohui) [reply wiring/src/spark_wiring_fuel.cpp:122]: That also makes sense. As `readRegister()` is private, we can change its prototype at our wish.

File: hal/src/nRF52840/i2c_hal.cpp (hunk 1 above)
Comment (XuGuohui) [reply hal/src/nRF52840/i2c_hal.cpp:394]: `HAL_I2C_Reset(i2c, 0, nullptr);` will simulates 9 clocks on SCL followed by calling `HAL_I2C_Begin()`, which will reinitialize I2C.

Comment (avtolstoy) [reply hal/src/nRF52840/i2c_hal.cpp:394]: :+1:

File: wiring/src/spark_wiring_fuel.cpp (hunk 2 above)
Comment (avtolstoy) [reply wiring/src/spark_wiring_fuel.cpp:122]: We can also do this as a separate PR as the main thing we are trying to resolve here is an unrecoverable (without manual intervention) bus lockup after an error.

**Pull request merged**